File: iosApp/iosApp/Modules/NFTDetail/NFTDetailWireframeFactory.swift
// Created by web3d4v on 27/05/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import UIKit
import web3walletcore

protocol NFTDetailWireframeFactory {
    func make(
        _ parent: UIViewController?,
        context: NFTDetailWireframeContext
    ) -> NFTDetailWireframe
}

final class DefaultNFTDetailWireframeFactory {
    private let nftSendWireframeFactory: NFTSendWireframeFactory
    private let alertWireframeFactory: AlertWireframeFactory
    private let nftsService: NFTsService
    private let walletService: WalletService
    private let networksService: NetworksService

    init(
        nftSendWireframeFactory: NFTSendWireframeFactory,
        alertWireframeFactory: AlertWireframeFactory,
        nftsService: NFTsService,
        walletService: WalletService,
        networksService: NetworksService
    ) {
        self.nftSendWireframeFactory = nftSendWireframeFactory
        self.alertWireframeFactory = alertWireframeFactory
        self.nftsService = nftsService
        self.walletService = walletService
        self.networksService = networksService
    }
}

extension DefaultNFTDetailWireframeFactory: NFTDetailWireframeFactory {

    func make(
        _ parent: UIViewController?,
        context: NFTDetailWireframeContext
    ) -> NFTDetailWireframe {
        DefaultNFTDetailWireframe(
            parent,
            context: context,
            nftSendWireframeFactory: nftSendWireframeFactory,
            alertWireframeFactory: alertWireframeFactory,
            nftsService: nftsService,
            walletService: walletService,
            networksService: networksService
        )
    }
}

// MARK: - Assembler

final class NFTDetailWireframeFactoryAssembler: AssemblerComponent {
    
    func register(to registry: AssemblerRegistry) {
        registry.register(scope: .instance) { resolver -> NFTDetailWireframeFactory in
            DefaultNFTDetailWireframeFactory(
                nftSendWireframeFactory: resolver.resolve(),
                alertWireframeFactory: resolver.resolve(),
                nftsService: resolver.resolve(),
                walletService: resolver.resolve(),
                networksService: resolver.resolve()
            )
        }
    }
}

File: androidApp/src/main/java/com/sonsofcrypto/web3wallet/android/modules/compose/nftdetail/NFTDetailWireframeFactory.kt
package com.sonsofcrypto.web3wallet.android.modules.compose.nftdetail

import androidx.fragment.app.Fragment
import com.sonsofcrypto.web3lib.services.networks.NetworksService
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3wallet.android.common.AssemblerComponent
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistry
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistryScope
import com.sonsofcrypto.web3wallet.android.modules.compose.nftsend.NFTSendWireframeFactory
import com.sonsofcrypto.web3walletcore.modules.nftDetail.NFTDetailWireframe
import com.sonsofcrypto.web3walletcore.modules.nftDetail.NFTDetailWireframeContext
import com.sonsofcrypto.web3walletcore.services.nfts.NFTsService
import smartadapter.internal.extension.name

interface NFTDetailWireframeFactory {
    fun make(parent: Fragment?, context: NFTDetailWireframeContext): NFTDetailWireframe
}

class DefaultNFTDetailWireframeFactory(
    private val nftsService: NFTsService,
    private val nftSendWireframeFactory: NFTSendWireframeFactory,
    private val networksService: NetworksService,
): NFTDetailWireframeFactory {

    override fun make(parent: Fragment?, context: NFTDetailWireframeContext): NFTDetailWireframe {
        return DefaultNFTDetailWireframe(
            parent?.let { WeakRef(it) },
            context,
            nftsService,
            nftSendWireframeFactory,
            networksService,
        )
    }
}

class NFTDetailWireframeFactoryAssembler: AssemblerComponent {

    override fun register(to: AssemblerRegistry) {

        to.register(NFTDetailWireframeFactory::class.name, AssemblerRegistryScope.INSTANCE) {

            DefaultNFTDetailWireframeFactory(
                it.resolve(NFTsService::class.name),
                it.resolve(NFTSendWireframeFactory::class.name),
                it.resolve(NetworksService::class.name),
            )
        }
    }
}